### DISCUSSION TITLE
2D/3D Acceleration

### DISCUSSION THREAD

**Discussion by mj-sakellaropoulos** [hansemro/kc1-linux#26] [General]
# Hardware
Imagination Graphics PowerVR SGX540

# Resources

- Reference manual page 2981 (Chapter 11)
- PVR DDK : https://powervr-graphics.github.io/
- Texas Instruments GIT : https://git.ti.com/cgit/graphics/
- Recent for beagleboneblack : https://elinux.org/BeagleBoneBlack/SGX_%2B_Qt_EGLFS_%2B_Weston
- OLD : https://android.googlesource.com/kernel/omap/+/794dd585c8e8dbe0c7fbdf682a74983929fe9147/drivers/gpu/pvr/Kconfig
- OLD : https://github.com/robclark/xf86-video-omap

### Driver
- Ongoing open source effort : https://github.com/openpvrsgx-devgroup/linux_openpvrsgx/blob/letux-pvrsrvkm-5.12-rc2/drivers/gpu/drm/pvrsgx
- 4430 : https://github.com/openpvrsgx-devgroup/linux_openpvrsgx/blob/letux-pvrsrvkm-5.12-rc2/drivers/gpu/drm/pvrsgx/1.9.2188537/eurasia_km/eurasiacon/build/linux2/omap4430_linux/Makefile
- Official driver ?

### X11 Acceleration Plugin
- Need PowerVR Driver and TI Kernel to match
- https://github.com/TexasInstruments/dri3wsegl

**Comment by hansemro** [comment]
https://github.com/openpvrsgx-devgroup/linux_openpvrsgx

https://wiki.postmarketos.org/wiki/Samsung_Galaxy_Tab_2_7.0%22_(samsung-espresso3g)

**Comment by hansemro** [comment]
https://gitlab.com/postmarketOS/pmaports/-/merge_requests/1868

**Comment by hansemro** [comment]
Of all Kindle hardware, this is what I am least familiar with.

**Reply by mj-sakellaropoulos** [reply]
Yeah. Graphics + ARM + Linux = Headache
It seems there are some recent efforts to streamline things, but its a long way off...

**Comment by hansemro** [comment]
repo: https://github.com/openpvrsgx-devgroup/linux_openpvrsgx
branch: `letux-pvrsrvkm-5.12-rc2`

<details>
<summary>Kernel config options</summary>
Preemptible Kernel (Low-latency Desktop): CONFIG_PREEMPT=y

PowerVR SGX: CONFIG_SGX=y
Using DDK 1.17.4948957
Leave CONFIG_SGX_DEBUG disabled
PowerVR SGX build with DRM support: CONFIG_SGX_DRM=y
PowerVR SGX inside OMAP SoC: CONFIG_SGX_OMAP=m
Leave CONFIG_SGX_GENERIC disabled
</details>

Packages:
```
sgx-ddk-um-ti443x
sgx-ddk-um-openrc
sgx-ddk-um
mesa-pvr-dri-classic
```

dmesg
```
[   16.170867] pvrsrvkm 56000000.gpu: Enabling quirks 00000001
[   16.205535] [drm] Initialized pvr 1.17.4948957 20110701 for 56000000.gpu on minor 1
```

<details>
<summary>SGXInitialize failure</summary>
<pre>
amazon-otter:~# lsmod
Module                  Size  Used by
wl12xx                 69632  0
wlcore                208896  1 wl12xx
mac80211              917504  2 wlcore,wl12xx
libarc4                16384  1 mac80211
cfg80211              847872  2 wlcore,mac80211
cpufreq_dt             16384  0
pvrsrvkm_omap4_sgx540_120   315392  0
wlcore_sdio            16384  0
evdev                  24576  2
mousedev               20480  0
omap_rng               16384  0
rng_core               20480  1 omap_rng
8250_omap              24576  0
8250                   40960  1 8250_omap
8250_base              49152  2 8250,8250_omap
serial_mctrl_gpio      16384  3 8250,8250_omap,8250_base
amazon-otter:~# pv
pvchange    pvcreate    pvmove      pvresize    pvs
pvck        pvdisplay   pvremove    pvrsrvinit  pvscan
amazon-otter:~# pvrsrvinit 
[   72.147216] PVR_K: UM DDK-(4948957) and KM DDK-(4948957) match. [ OK ]
[   73.195678] PVR_K:(Error): PollForValueKM: Timeout. Expected 0x1 but found 0x0 (mask 0x1).
[   73.204895] PVR_K:(Error): SGXInitialise: Wait for uKernel initialisation failed
[   73.213226] PVR_K: SGX debug (SGX_DDK sgxddk 1.17@4948957)
[   73.219207] PVR_K:  Host Ctl flags= 00000004
[   73.223785] PVR_K: SGX Host control:
[   73.227874] PVR_K: 	(HC-0) 0x00000000 0x00000000 0x00000000 0x00000000
[   73.234771] PVR_K: 	(HC-10) 0x00000000 0x00000014 0x00030D40 0x00000003
[   73.242095] PVR_K: 	(HC-20) 0x00000000 0x00000000 0x00000000 0x00000000
[   73.249237] PVR_K: 	(HC-30) 0x00000000 0xF26BE530 0x00000000 0x00000000
[   73.256347] PVR_K: 	(HC-40) 0x00000000 0x00000000 0x616D7609 0x732E7478
[   73.263336] PVR_K: SGX TA/3D control:
[   73.267333] PVR_K: 	(T3C-0) 0x0F003000 0x0F003120 0x0F002000 0x00000000
[   73.274383] PVR_K: 	(T3C-10) 0x00000000 0x00000002 0x00000000 0x00000000
[   73.281738] PVR_K: 	(T3C-20) 0x00000000 0x00000000 0x00000000 0x00000000
[   73.289031] PVR_K: 	(T3C-30) 0x00000000 0x00000000 0x00000000 0x00000000
[   73.296813] PVR_K: 	(T3C-40) 0x00000000 0x00000000 0x00000000 0x00000000
[   73.303924] PVR_K: 	(T3C-50) 0x00000000 0x00000000 0x00000000 0x00000000
[   73.311431] PVR_K: 	(T3C-60) 0x00000000 0x00000000 0x00000000 0x00000000
[   73.318664] PVR_K: 	(T3C-70) 0x00000000 0x00000000 0x00000000 0x00000000
[   73.325866] PVR_K: 	(T3C-80) 0x00000000 0x00000000 0x0F000000 0x84EEA000
[   73.332977] PVR_K: 	(T3C-90) 0x00000000 0x00000000 0x00000000 0x00000000
[   73.340209] PVR_K: 	(T3C-A0) 0x00000000 0x00000000 0x00000000 0x00000000
[   73.347686] PVR_K: 	(T3C-B0) 0x00000000 0x00000000 0x00000000 0x00000000
[   73.354827] PVR_K: 	(T3C-C0) 0x00000000 0x00000000 0x00000000 0x0F004000
[   73.375427] PVR_K: 	(T3C-D0) 0x0F00A420 0x0F00B000 0x00000000 0x00000000
[   73.395538] PVR_K: 	(T3C-E0) 0x00000000 0x00000000 0x00000000 0x00000000
[   73.402648] PVR_K: 	(T3C-F0) 0x00000000 0x00000000 0x00000000 0x00000000
[   73.425384] PVR_K: 	(T3C-100) 0x00000000 0x00000000 0x00000000 0x00000000
[   73.432586] PVR_K: 	(T3C-110) 0x00000000 0x00000000 0x25202C32 0x6D003371
[   73.439971] PVR_K: SGX Kernel CCB WO:0x0 RO:0x0
[   73.444763] PVR_K: Active syncs
[   73.448425] PVR_K:(Error): SGXPostPowerState: SGXInitialise failed
[   73.455749] PVR_K:(Error): PVRSRVSetDevicePowerStateKM : Transition to 0 FAILED 0x18
[   73.463958] PVR_K:(Error): SGXScheduleCCBCommandKM failed to power up device - ui32CallerID:-1 eError:24
[   73.474426] PVR_K:(Error): SGXGetMiscInfoUkernel: SGXScheduleCCBCommandKM failed.
[   73.482482] PVR_K: (FAIL) SGXInit: Unable to validate hardware core revision
[   73.490081] PVR_K:(Error): PVRSRVFinaliseSystem: Failed PVRSRVDevInitCompatCheck call (device index: 0)
PVR:(Error): PVRSRVInitSrvDisconnect: KM returne[   73.500671] PVR_K:(Error): BridgedDispatchKM: Initialisation failed.  Driver unusable.
d 24 [0, ]
PVR:(Error): PVRSRVBridgeCall: Failed to access device.  Function ID:3223086862 (strerror returns no value.). [0, ]
PVR:(Error): SrvInit: PVRSRVInitSrvDisconnect failed (24). See srvkm log for details. [0, ]
failed to initialize server
</pre>
</details>

**Reply by mj-sakellaropoulos** [reply]
From the logs, I'm wondering if it's another case of missing voltage regulator / clock / both ?

**Reply by MightyM17** [reply]
https://github.com/openpvrsgx-devgroup/linux_openpvrsgx/issues/3#issuecomment-803759547 fix is here

**Reply by hansemro** [reply]
Adding `-DPVR_LINUX_MEM_AREA_USE_VMAP` to pvrsgx Makefile (like in [here](https://github.com/MightyM17/linux_openpvrsgx/commit/7d75452c82b973c956db003de0336e8c48267d0c#diff-62c5cc773d1a0e3287b4008ee26ada3574de7bb9f3b637406cd1d8d40be8b003)) fixes the driver enough to run Wayland and some demos. Weston and kmscube have gotten smoother.

<details>
<summary>60fps kmscube with PVRSGX DDK 1.17.4948957 (letux-pvrsrvkm-5.13-rc2)</summary>
<pre>
amazon-otter:~# kmscube 
Using display 0xb6dfa0a0 with EGL version 1.4
===================================
EGL information:
  version: "1.4"
  vendor: "Mesa Project"
  client extensions: "EGL_EXT_client_extensions EGL_EXT_device_base EGL_EXT_device_enumeration EGL_EXT_device_query EGL_EXT_platform_base EGL_KHR_client_get_all_proc_addresses EGL_KHR_debug EGL_EXT_platform_device EGL_EXT_platform_wayland EGL_KHR_platform_wayland EGL_EXT_platform_x11 EGL_KHR_platform_x11 EGL_MESA_platform_xcb EGL_MESA_platform_gbm EGL_KHR_platform_gbm EGL_MESA_platform_surfaceless"
  display extensions: "EGL_EXT_buffer_age EGL_EXT_create_context_robustness EGL_EXT_image_dma_buf_import EGL_EXT_image_dma_buf_import_modifiers EGL_IMG_context_priority EGL_KHR_config_attribs EGL_KHR_create_context EGL_KHR_fence_sync EGL_KHR_get_all_proc_addresses EGL_KHR_gl_renderbuffer_image EGL_KHR_gl_texture_2D_image EGL_KHR_gl_texture_cubemap_image EGL_KHR_image EGL_KHR_image_base EGL_KHR_image_pixmap EGL_KHR_no_config_context EGL_KHR_reusable_sync EGL_KHR_surfaceless_context EGL_EXT_pixel_format_float EGL_KHR_wait_sync EGL_MESA_configless_context EGL_MESA_drm_image EGL_WL_bind_wayland_display "
===================================
OpenGL ES 2.x information:
  version: "OpenGL ES 2.0 build 1.17@4948957"
  shading language version: "OpenGL ES GLSL ES 1.00 build 1.17@4948957"
  vendor: "Imagination Technologies"
  renderer: "PowerVR SGX 540"
  extensions: "GL_OES_compressed_ETC1_RGB8_texture GL_OES_depth24 GL_OES_depth_texture GL_OES_egl_sync GL_OES_element_index_uint GL_OES_EGL_image GL_OES_EGL_image_external GL_OES_fbo_render_mipmap GL_OES_fragment_precision_high GL_OES_get_program_binary GL_OES_mapbuffer GL_OES_packed_depth_stencil GL_OES_required_internalformat GL_OES_rgb8_rgba8 GL_OES_standard_derivatives GL_OES_surfaceless_context GL_OES_texture_float GL_OES_texture_half_float GL_OES_vertex_array_object GL_OES_vertex_half_float GL_EXT_blend_minmax GL_EXT_discard_framebuffer GL_EXT_multi_draw_arrays GL_EXT_multisampled_render_to_texture GL_EXT_shader_texture_lod GL_EXT_texture_format_BGRA8888 GL_EXT_texture_rg GL_IMG_multisampled_render_to_texture GL_IMG_program_binary GL_IMG_read_format GL_IMG_shader_binary GL_IMG_texture_compression_pvrtc GL_IMG_texture_format_BGRA8888 GL_IMG_texture_npot GL_IMG_uniform_buffer_object GL_KHR_debug GL_EXT_texture_storage"
===================================
Using modifier ffffffffffffff
Modifiers failed!
Using modifier ffffffffffffff
Modifiers failed!
Rendered 120 frames in 2.001244 sec (59.962700 fps)
Rendered 240 frames in 4.001329 sec (59.980078 fps)
Rendered 360 frames in 6.001779 sec (59.982213 fps)
Rendered 480 frames in 8.001864 sec (59.986026 fps)
Rendered 600 frames in 10.002253 sec (59.986484 fps)
</pre>
</details>
